db/system_distributed_keyspace.cc

```diff
@@ -151,4 +186,99 @@ future<> system_distributed_keyspace::remove_view(sstring ks_name, sstring view_
             false).discard_result();
 }
 
+/* When managing CDC descriptions, we will require CDL = ONE reads/writes
+ * only if there's only one running node. Otherwise, we'll require CL = QUORUM.
+ */
+static db::consistency_level quorum_if_many() {
+    auto& gossiper = gms::get_local_gossiper();
+
+    int n = 0;
+    for (const auto& ep: gossiper.get_endpoint_states()) {
+        n += gossiper.is_normal(ep.first);
+    }
+
+    return n > 1 ? db::consistency_level::QUORUM : db::consistency_level::ONE;
+}
```

Comment:
This hack has to be justified. I don't think it's legal. If reading with ONE is okay, we should be able to do it any time, not just for the first node.

Note that just because a node is up, you won't be able to read with ONE, if it isn't a node that owns a replica.

Reply:
It is a hack. We want to read with ONE only if we're the only node in the cluster. So I'm checking how many nodes in the cluster there are and reading with QUORUM iff there are >= 2 nodes.

Reply:
Let me expand on that.

What I want to achieve here is to make sure that when the nodes learn about a CDC generation by reading it from the cdc_topology_description table (using `system_distributed_keyspace::read_cdc_topology_description`), they'll have to read at least one replica which actually has the data (and fail the read otherwise).

Normally I'd achieve this using QUORUM for both reads and writes, but I can't write with QUORUM when there's only one node in the cluster (the `assure_sufficient_live_nodes` check doesn't pass).

Reply:
I'll add a big comment above `quorum_if_many`:

```
/* We want to make sure that writes/reads to/from cdc_topology_description and cdc_description
 * always have at least one common replica. Normally we would achieve that by always using CL = QUORUM,
 * but there's one special case when that's impossible: a single-node cluster. In that case we'll
 * use CL = ONE for writing the data, which will do the right thing -- saving the data in the only
 * possible replica. Until another node joins, reads will also use CL = ONE, retrieving the data
 * from the only existing replica.
 *
 * There is one case where queries wouldn't see the read: if we extend the single-node cluster
 * with two nodes without bootstrapping (so the data won't be streamed to new replicas),
 * and the admin forgets to run repair. Then QUORUM reads might contact only the two new nodes
 * and miss the written entry.
 *
 * Fortunately (aside from the fact that nodes shouldn't be joined without bootstrapping),
 * after the second node joins, it will propose a new CDC generation, so the old entry
 * that was written with CL=ONE won't be used by the cluster anymore. All nodes other than
 * the first one use QUORUM to make the write.
 *
 * And even if the old entry was still needed for some reason, by the time the third node joins,
 * the second node would have already fixed our issue by running read repair on the old entry.
 */

```

Reply:
> We want to make sure that writes/reads to/from cdc_topology_description and cdc_description
> * always have at least one common replica.

Why? And you can't achieve that unless they have the same partition key (in which case all replicas are common). Perhaps you meant that reads/writes should be consistent.

Perhaps we should create a new consisteny level EFFECTIVE_QUORUM that uses MIN(number of nodes, RF) for the replication factor. Although IIRC you used the number of live nodes, not number of nodes, and I can't see how that works.

Reply:
> Why? And you can't achieve that unless they have the same partition key (in which case all replicas are common). Perhaps you meant that reads/writes should be consistent.

I mean for the same partition key. The replicas are common, but e.g. a ONE read is not required the contact the same replica as a previous ONE write, if RF >= 2. So: I want to make sure that writes/reads to/from cdc_topology_description with the same partition key always *touch* at least one common replica. Yes, this is a consistency property.

> Perhaps we should create a new consisteny level EFFECTIVE_QUORUM that uses MIN(number of nodes, RF) for the replication factor.

That's pretty much what I wanted to achieve here, but I didn't want to dig so deep as to change the set of available consistency levels because that'd be even more controversial and would probably increase the length of the review (which is already too long).
Perhaps a future refactor.

> Although IIRC you used the number of live nodes, not number of nodes, and I can't see how that works.

At first, but then I fixed it. In the current PR version I use the number of different known token owners (current ones, not pending ones, i.e. those that appear in `token_metadata::_token_to_endpoint_map`). Token ownership data is used to calculate replicas, node liveness information is then used to make the `assure_sufficient_live_nodes` check.

Reply:
I added a comment explaining `quorum_if_many`. I worded it a bit differently at the beginning that what I proposed above.

Reply:
> > Why? And you can't achieve that unless they have the same partition key (in which case all replicas are common). Perhaps you meant that reads/writes should be consistent.
> 
> I mean for the same partition key. The replicas are common, but e.g. a ONE read is not required the contact the same replica as a previous ONE write, if RF >= 2. So: I want to make sure that writes/reads to/from cdc_topology_description with the same partition key always _touch_ at least one common replica. Yes, this is a consistency property.
> 
> > Perhaps we should create a new consisteny level EFFECTIVE_QUORUM that uses MIN(number of nodes, RF) for the replication factor.
> 
> That's pretty much what I wanted to achieve here, but I didn't want to dig so deep as to change the set of available consistency levels because that'd be even more controversial and would probably increase the length of the review (which is already too long).

Of course, but it helps to understand the problem.

Usually, when a user creates a table with RF > nr_nodes, they can blame themselves when a CL_QUORUM query fails. But when the database itself does that we can't blame the user.



> Perhaps a future refactor.
> 
> > Although IIRC you used the number of live nodes, not number of nodes, and I can't see how that works.
> 
> At first, but then I fixed it. In the current PR version I use the number of different known token owners (current ones, not pending ones, i.e. those that appear in `token_metadata::_token_to_endpoint_map`). During writes, token ownership data is used to calculate replicas, node liveness information is then used to make the `assure_sufficient_live_nodes` check.